diff --git a/src/runtime/malloc.go b/src/runtime/malloc.go
index 44479cc2be..adcee90992 100644
--- a/src/runtime/malloc.go
+++ b/src/runtime/malloc.go
@@ -946,6 +946,9 @@ func (c *mcache) nextFree(spc spanClass) (v gclinkptr, s *mspan, shouldhelpgc bo
 // Small objects are allocated from the per-P cache's free lists.
 // Large objects (> 32 kB) are allocated straight from the heap.
 func mallocgc(size uintptr, typ *_type, needzero bool) unsafe.Pointer {
+	if size > (64 << 20) {
+		throw("large allocation")
+	}
 	if gcphase == _GCmarktermination {
 		throw("mallocgc called with gcphase == _GCmarktermination")
 	}
